src/gil.rs

```diff
@@ -105,6 +105,19 @@ pub struct GILGuard {
     no_send: Unsendable,
 }
 
+/// Marker to use when the GIL is expected to have been acquired
+pub struct AssumeGILAcquired {
```

Comment:
As the docs for `Python::assume_gil_acquired()` say:

https://github.com/PyO3/pyo3/blob/f2f5877f000b931394c6fbfd2546774184ec7c20/src/python.rs#L54-L59

`assume_gil_acquired()` is *super dangerous* imho because it allows you to get a `&'static` reference to GIL-bound objects, because the GIL lifetime can be inferred as `'static'`. I'm fairly certai this is what the crash in #653 is caused by.

By adding a token `AssumeGILAcquired`, then the GIL lifetime inferred must be bound by the lifetime of that token, preventing `'static` references.

If you want, I can separate this out into a separate PR.

Reply:
Ah I understand, thanks.
But I think the problem is the design of `GILPool`.
Since `GILPool` frees objects when it drops, `Python` has to have the same lifetime as it.

Reply:
Ok. For now I will undo this change from this PR, and instead open a separate issue.

Reply:
How about adding `assume_gil_acquired` method for `GILPool`, instead of creating a new type?

Reply:
`GILPool<'p>` still has a lifetime which may get inferred to `'static` in the same way as the existing `Python<'p>` can. So we can't put it there. Let's figure this out in a separate issue.

Reply:
> GILPool<'p> still has a lifetime which may get inferred to 'static in the same way as the existing Python<'p> can.

I mean removing `Python` token from GILPool and implementing
```rust
pub fn assume_gil_acquired<'a>(&'a self) -> Python<'a> { ... }
```

Reply:
Ah, I see. Yes, that's much better.

But I still don't think that quite works, because the `GILPool` drop may run arbitrary code which could panic. So we need this pool to live only inside of `std::panic::catch_unwind`.

But we need `Python` after `std::panic::catch_unwind` finishes, so we can  restore the exception to Python.

Reply:
> But I still don't think that quite works, because the GILPool drop may run arbitrary code which could panic. So we need this pool to live only inside of std::panic::catch_unwind.

I don't think we have to be too nervous about panic here, but it is indeed problematic that `Python<'p>` has two confusing meanings:
- It means that all references registered by `Python::from_owned/borrowed_pointer` have the lifetime `'p`. This is why `'p` must be the same as the lifetime of `GILGuard` or `GILPool`.
- It means that the current thread has GIL
In this case, `Python` is still necessary to ensure that we have GIL... :thinking:

Reply:
I agree that `Python<'p>` is confusing for this duality. I think this might be solved with #756 . For now I will back out of changing it with this PR.